spec/notification_center_spec.rb

```diff
@@ -0,0 +1,203 @@
+
+      it 'should log and return nil when notification type is invalid' do
+        notification_center = Optimizely::NotificationCenter.new(spy_logger)
+        expect(notification_center.add_notification_listener(
+                 Optimizely::NotificationCenter::NOTIFICATION_TYPES[:DECISION],
+                 'Invalid callback!'
+        )).to eq(nil)
+        expect(spy_logger).to have_received(:log).once
+          .with(Logger::ERROR, 'Invalid callback! is invalid.')
+      end
+
+      it 'should return 1 for valid params' do
+        notification_center = Optimizely::NotificationCenter.new(Optimizely::SimpleLogger.new)
+        expect(notification_center.add_notification_listener(
+                 Optimizely::NotificationCenter::NOTIFICATION_TYPES[:DECISION],
+                 @callback_reference
+        )).to eq(1)
+      end
+
+      it 'shoud return -1 when callback already exists' do
+        notification_center = Optimizely::NotificationCenter.new(Optimizely::SimpleLogger.new)
+        notification_center.add_notification_listener(
+          Optimizely::NotificationCenter::NOTIFICATION_TYPES[:DECISION],
+          @callback_reference
+        )
+        expect(notification_center.add_notification_listener(
+                 Optimizely::NotificationCenter::NOTIFICATION_TYPES[:DECISION],
+                 @callback_reference
+        )).to eq(-1)
+      end
+    end
+
+    describe '.remove_notification_listener' do
+      before(:example) do
+        @notification_type = Optimizely::NotificationCenter::NOTIFICATION_TYPES[:DECISION]
+        @notification_center = Optimizely::NotificationCenter.new(spy_logger)
+        @notification_center.add_notification_listener(@notification_type, @callback_reference)
+      end
+
+      it 'shoud log and return nil when notification id is blank' do
+        expect(@notification_center.remove_notification_listener(nil)).to eq(nil)
+        expect(spy_logger).to have_received(:log).once
+          .with(Logger::ERROR, 'Notification id can not be blank!')
+      end
+
+      it 'shoud return true if notification is removed' do
+        expect(@notification_center.remove_notification_listener(1)).to eq(true)
+      end
+
+      it 'shoud return false if notification failed to remove' do
+        expect(@notification_center.remove_notification_listener(2))
+          .to eq(false)
+      end
+    end
+
+    describe '.clear_notifications' do
+      it 'should return log of notifications cleared' do
+        notification_type = Optimizely::NotificationCenter::NOTIFICATION_TYPES[:DECISION]
+        notification_center = Optimizely::NotificationCenter.new(spy_logger)
+        notification_center.add_notification_listener(notification_type, @callback_reference)
+        notification_center.clear_notifications(notification_type)
+        expect(spy_logger).to have_received(:log).once
+          .with(Logger::INFO, "All callbacks for notification type #{notification_type} have been removed.")
```

Comment:
See PHP SDK for reference. 
Check clear notifications by adding multiple notifications for different types. 
And assert that Notifications for Type B remains intact when all for type A have been removed.

Reply:
Done